### ISSUE TITLE
Time Wrong in the cloud

### ISSUE THREAD

**Issue by eXorus** [Codeception/Codeception#3262]
Hello,

I'm running 6 tests on local with docker and selenium grid :
```
$ ./vendor/bin/codecept run --env dockerLocal 
Codeception PHP Testing Framework v2.2.1
Powered by PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Acceptance (dockerLocal) Tests (6) ----------------------------------------------------------------------------------------------
✔ testCest: Try to failed activate with wrong code (31.3107s)
✔ testCest: Try to failed activate with names (52.5252s)
✔ testCest: Try to failed activate with last name (47.4771s)
✔ testCest: Try to activate successfully (46.4682s)
✔ testCest: Try to failed activate (52.5289s)
✔ testCest: Try to failed activate with first name (53.5311s)
---------------------------------------------------------------------------------------------------------------------------------
Time: 4.95 minutes, Memory: 19.50MB
OK (6 tests, 193 assertions)
```

Now if I'm running the same test with SauceLabs :
```
$ ./vendor/bin/codecept run --env saucelabs  
Codeception PHP Testing Framework v2.2.1
Powered by PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Acceptance (saucelabs) Tests (6) ------------------------------------------------------------------------------------------------
✔ testCest: Try to failed activate with names (14.25421s)
✔ testCest: Try to failed activate (12.25245s)
✔ testCest: Try to failed activate with wrong code (13.7351s)
✔ testCest: Try to failed activate with last name (55.23544s)
✔ testCest: Try to activate successfully (10.13018s)
✔ testCest: Try to failed activate with first name (10.25072s)
---------------------------------------------------------------------------------------------------------------------------------
Time: 20.56 minutes, Memory: 19.50MB
OK (6 tests, 193 assertions)
```

How is it possible that 14s + 12s + 13s + 55s + 10s + 10s = 20 minutes ?

**Comment by Naktibalda** [comment]
How long did it actually take?

**Comment by eXorus** [comment]
20 minutes really

**Comment by eXorus** [comment]
I think it's issue in PHPUnit because the diff is between local date and remote date... but I'm not sure

**Comment by Naktibalda** [comment]
The execution time of individual test is measured by [Codeception](https://github.com/Codeception/Codeception/blob/e31097051056e26d2d70e23b4df57a0f6f7df1b0/src/Codeception/Test/Test.php#L80) and I believe that overall time and memory usage is measured by PhpUnit.

Execution time of before and after hooks is not included to execution time of individual tests.
Is it possible that setup or cleanup could take that long?

**Comment by eXorus** [comment]
there is nothing in _before() or _after()

**Comment by Naktibalda** [comment]
Codeception modules have their own before and after.

**Comment by eXorus** [comment]
Where ? I can add var_dump to trace the time :)

**Comment by Naktibalda** [comment]
https://github.com/Codeception/Codeception/blob/2.2/src/Codeception/Module/WebDriver.php#L314
https://github.com/Codeception/Codeception/blob/2.2/src/Codeception/Module/Db.php#L221

**Comment by eXorus** [comment]
Thanks, this is the result more focus on one test


It take really 2 min but Codeception said 3 seconds :
```
$ ./vendor/bin/codecept run --env saucelabs
Codeception PHP Testing Framework v2.2.1
Powered by PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Acceptance (saucelabs) Tests (1) ----------------------------------------------------------------------------------------------------------------
✔ testCest: test (3.12349s)
-------------------------------------------------------------------------------------------------------------------------------------------------
Time: 2.15 minutes, Memory: 17.75MB
OK (1 test, 21 assertions)
```

Now in debug with var_dump(DATE_ATOM) in Test.php, Db.php and WebDriver.php :
```
$ ./vendor/bin/codecept run --debug --env saucelabs
Codeception PHP Testing Framework v2.2.1
Powered by PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Acceptance (saucelabs) Tests (1) ----------------------------------------------------------------------------------------------------------------
Modules: Cli, WebDriver, \Helper\Acceptance, \Helper\Api
-------------------------------------------------------------------------------------------------------------------------------------------------
testCest: test
Signature: testCest:test
Test: tests/acceptance/testCest.php:test
Scenario --
string(54) "WebDriver.php Before Start : 2016-06-22T18:14:33+02:00"
string(52) "WebDriver.php Before End : 2016-06-22T18:14:33+02:00"
string(42) "Test.php Start : 2016-06-22T18:14:33+02:00"
 I am on page "/xxx"
  [GET] https://xxx
 I fill field "LOGIN","xxx"
 I fill field "CODEIW","xxxx"
...
after 30 steps
...
 USER DOES --- 
 I am on page "/yyy"
  [GET] https://yyy
 I wait for text "Start",60
 I click "Start"
...
 after 50 steps
...
 --- USER FINISHED
string(41) "Test.php Stop : 2016-06-22T18:16:36+02:00"
 PASSED 

-------------------------------------------------------------------------------------------------------------------------------------------------
Time: 2.15 minutes, Memory: 17.75MB
OK (1 test, 21 assertions)
```

**Comment by eXorus** [comment]
Without debug :
```
$ ./vendor/bin/codecept run --env saucelabs 
Codeception PHP Testing Framework v2.2.1
Powered by PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Acceptance (saucelabs) Tests (1) ----------------------------------------------------------------------------------------------------------------
- testCest: test
string(54) "WebDriver.php Before Start : 2016-06-22T18:39:49+02:00"
string(52) "WebDriver.php Before End : 2016-06-22T18:39:49+02:00"
string(42) "Test.php Start : 2016-06-22T18:39:49+02:00"
string(41) "Test.php Stop : 2016-06-22T18:41:59+02:00"
✔ testCest: test (10.13004s)
-------------------------------------------------------------------------------------------------------------------------------------------------
Time: 2.26 minutes, Memory: 17.75MB
OK (1 test, 21 assertions)
```

**Comment by Naktibalda** [comment]
I've looked at the source of [PHP_Timer](https://github.com/sebastianbergmann/php-timer/blob/master/src/Timer.php#L40). It looks that it could return unexpected results if multiple timers are used at the same time and are not ended in correct order.
```
start timer #1;
sleep 10;
start timer #2:
sleep 1;
stop timer #1; //result 1
sleep 1;
stop timer #2; // result 12;
```

Could you do some debugging and check how many timers are actually used and what results they return?

**Comment by eXorus** [comment]
I added var_dump(self::$startTimes)
at the end of start function in Timer.php
and at the top of the stop function in Timer.php

This is the result :
```
$ ./vendor/bin/codecept run --env saucelabs
Codeception PHP Testing Framework v2.2.1
Powered by PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Acceptance (saucelabs) Tests (1) ----------------------------------------------------------------------------------------------------------------
- testCest: test
string(54) "WebDriver.php Before Start : 2016-06-22T19:00:36+02:00"
string(52) "WebDriver.php Before End : 2016-06-22T19:00:36+02:00"
string(42) "Test.php Start : 2016-06-22T19:00:36+02:00"
array(1) {
  [0]=>
  float(1466614836.9167)
}
array(1) {
  [0]=>
  float(1466614836.9167)
}
string(41) "Test.php Stop : 2016-06-22T19:02:42+02:00"
✔ testCest: test (5.12519s)
-------------------------------------------------------------------------------------------------------------------------------------------------
Time: 2.18 minutes, Memory: 17.75MB
OK (1 test, 21 assertions)
```

**Comment by Naktibalda** [comment]
I did some debugging and now I think that the extra time is not taken by the gaps, but it is mostly preparation time before starting the first test and report generation after finishing the last test.

`var_dump(\PHP_Timer::$requestTime); `

**Comment by eXorus** [comment]
I don't think so, I added this 4 lines at different point of my test function

```php
var_dump("requestTime : ");
var_dump(\PHP_Timer::$requestTime);
var_dump("timeSinceStartOfRequest : ");
var_dump(\PHP_Timer::timeSinceStartOfRequest());
```

And I added "#### I begin my test" inside my test function at the first line, and "#### I finish my test" inside my test function at the end.

This is the result, so we see that time is taken by the step and not the preparation or the report generation. How adn Where do you calculate the (1.12159s) because if you take timeSinceStartOfRequest() it will be ok.


```
$ ./vendor/bin/codecept run --env saucelabs
Codeception PHP Testing Framework v2.2.1
Powered by PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Acceptance (saucelabs) Tests (1) ------------------------------------------------------------------------------------------------------------------------------
- testCest: test
string(54) "WebDriver.php Before Start : 2016-06-23T09:17:33+02:00"
string(52) "WebDriver.php Before End : 2016-06-23T09:17:33+02:00"
string(42) "Test.php Start : 2016-06-23T09:17:33+02:00"
string(20) "#### I begin my test"
string(14) "requestTime : "
float(1466666248.4128)
string(26) "timeSinceStartOfRequest : "
string(11) "5.4 seconds"
string(14) "requestTime : "
float(1466666248.4128)
string(26) "timeSinceStartOfRequest : "
string(12) "5.42 seconds"
string(14) "requestTime : "
float(1466666248.4128)
string(26) "timeSinceStartOfRequest : "
string(12) "51.2 seconds"
string(14) "requestTime : "
float(1466666248.4128)
string(26) "timeSinceStartOfRequest : "
string(13) "56.43 seconds"
string(14) "requestTime : "
float(1466666248.4128)
string(26) "timeSinceStartOfRequest : "
string(12) "2.11 minutes"
string(14) "requestTime : "
float(1466666248.4128)
string(26) "timeSinceStartOfRequest : "
string(12) "2.11 minutes"
string(21) "#### I finish my test"
string(41) "Test.php Stop : 2016-06-23T09:19:35+02:00"
✔ testCest: test (1.12159s)
---------------------------------------------------------------------------------------------------------------------------------------------------------------


Time: 2.12 minutes, Memory: 17.75MB

OK (1 test, 21 assertions)

```

**Comment by Naktibalda** [comment]
The execution time of individual tests is measured by  https://github.com/Codeception/Codeception/blob/e31097051056e26d2d70e23b4df57a0f6f7df1b0/src/Codeception/Test/Test.php#L80

**Comment by eXorus** [comment]
I was trying to find where do you print (xxx,xxxs) ?

Because $time return by stop() is right so there is something wrong after.

```
string(15) "Stop Time is : "
float(123.75530815125)
string(10) "In TestResult.php time is"
float(123.75530815125)
✔ virtualAuthenticatorCest: Try to activate successfully (3.12375s)
```

The endTest is called with 123 sec and the print is 3 sec ????
 $result->endTest($this, $time);

**Comment by DavertMik** [comment]
Should be fixed by #3365

**Comment by eXorus** [comment]
**Codeception PHP Testing Framework v2.2.2**
```
Powered by PHPUnit 5.4.6 by Sebastian Bergmann and contributors.

Acceptance (testTime) Tests (6) 

✔ virtualAuthenticatorCest: Try to failed activate with names (12.13264s)
✔ virtualAuthenticatorCest: Try to activate successfully (22.8228s)
✔ virtualAuthenticatorCest: Try to failed activate with first name (8.12845s)
✔ virtualAuthenticatorCest: Try to failed activate with last name (41.10105s)
✔ virtualAuthenticatorCest: Try to failed activate (4.12454s)
✔ virtualAuthenticatorCest: Try to failed activate va with wrong code (35.3599s)

Time: 10.85 minutes, Memory: 20.50MB
```

**Codeception PHP Testing Framework v2.2.3**
```
Powered by PHPUnit 5.4.7 by Sebastian Bergmann and contributors.

Acceptance (testTime) Tests (6) 

✔ virtualAuthenticatorCest: Try to failed activate with names (145.69s)
✔ virtualAuthenticatorCest: Try to activate successfully (88.46s)
✔ virtualAuthenticatorCest: Try to failed activate va with wrong code (38.08s)
✔ virtualAuthenticatorCest: Try to failed activate with first name (131.68s)
✔ virtualAuthenticatorCest: Try to failed activate with last name (96.82s)
✔ virtualAuthenticatorCest: Try to failed activate (135.50s)

Time: 11.33 minutes, Memory: 20.00MB

```

Confirmed !!! Thanks @DavertMik
